faker-js/faker
# feat: migrate locales to typescript

I automatically transformed the locales using the following `java` script.

<details><summary>Source (Click to expand)</summary>

* Save this as Convert.java inside your `Faker` checkout.
* Run `java Convert.java` using java 17

````java
import static java.nio.charset.StandardCharsets.UTF_8;

import java.io.BufferedWriter;
import java.io.File;
import java.io.FileWriter;
import java.io.IOException;
import java.nio.file.Files;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;
import java.util.TreeSet;
import java.util.stream.Collectors;

public class Convert {

    static final File root = new File("src/locales");

    public static void main(final String... args) throws IOException {
        transform(root, 0);
    }

    private static void transform(final File file, final int depth) throws IOException {
        final String name = file.getName();
        if (!file.exists() || name.equals("index.js") || name.equals("index.ts")) {
            return;
        }
        if (file.isDirectory()) {
            for (final File sub : file.listFiles()) {
                transform(sub, depth + 1);
            }
            buildIndex(file, depth);
        } else if (name.endsWith(".ts")) {
            transformTs(file);
        } else if (name.endsWith(".js")) {
            renameJs(file);
        }
    }

    private static void buildIndex(final File folder, final int depth) throws IOException {
        final TreeSet<String> files = Arrays.stream(folder.listFiles())
                .map(File::getName)
                .map(s -> s.replace(".ts", ""))
                .filter(f -> !f.startsWith("index"))
                .collect(Collectors.toCollection(TreeSet::new));

        final String name = folder.getName();


        final File index = new File(folder, "index.ts");
        Map<String, String> indexKeys = new HashMap<>();
        if (index.exists() && depth == 1) {
            indexKeys = Files.lines(index.toPath(), UTF_8)
                    .map(String::trim)
                    .map(s -> s.split(":"))
                    .filter(s -> s.length == 2)
                    // .peek(s -> System.out.println(s[0] + ": " + s[1]))
                    .collect(Collectors.toMap(s -> s[0], s -> s[1]));
        }

        try (FileWriter fw = new FileWriter(new File(folder, "index.ts"), UTF_8);
                BufferedWriter writer = new BufferedWriter(fw)) {
            if (depth == 1) {

                System.out.println("------------------------------------------------");
                System.out.println(name + "/index.ts");

                fw.write("import type { LocaleDefinition } from '../..';\n");
                for (final String file : files) {
                    fw.write("import " + file + " from './" + file + "';\n");
                }
                fw.write("\n");
                fw.write("const " + name + ": LocaleDefinition = {\n");
                fw.write("  title:" + indexKeys.get("title") + "\n");
                if (indexKeys.containsKey("separator")) {
                    fw.write("  separator:" + indexKeys.get("separator") + "\n");
                }
                for (final String file : files) {
                    fw.write("  " + file + ",\n");
                }
                fw.write("};\n");
                fw.write("\n");
                fw.write("export default " + name + ";\n");
            } else {
                final String importPath = ".." + ("/..".repeat(depth));
                fw.write("import type { LocaleDefinition } from '" + importPath + "';\n");
                for (final String file : files) {
                    if (file.equals("switch") || file.equals(name)) {
                        fw.write("import " + file + "_ from './" + file + "';\n");
                    } else {
                        fw.write("import " + file + " from './" + file + "';\n");
                    }
                }
                fw.write("\n");
                final String type = extractType(folder, depth);
                fw.write("const " + name + ": " + type + " = {\n");
                for (final String file : files) {
                    if (file.equals("switch") || file.equals(name)) {
                        fw.write("  " + file + ": " + file + "_,\n");
                    } else {
                        fw.write("  " + file + ",\n");
                    }
                }
                fw.write("};\n");
                fw.write("\n");
                fw.write("export default " + name + ";\n");
            }
        }
        final File indexJs = new File(folder, "index.js");
        if (indexJs.exists()) {
            indexJs.delete();
        }
    }

    private static String extractType(final File folder, final int depth) {
        if (depth == 0) {
            return "{ [lang: string]: LocaleDefinition }";
        }
        final String[] subPath =
                folder.getPath().substring(root.getPath().length() + 1).replace('\\', '/').split("/");
        final String type = "LocaleDefinition" + Arrays.stream(subPath)
                .skip(1).map(s -> "['" + s + "']")
                .collect(Collectors.joining());
        return type;
    }

    private static void transformTs(final File file) throws IOException {
        String content = Files.readString(file.toPath(), UTF_8);
        content = content.replace("module['exports'] =", "export default");
        content = content.replace("module.exports =", "export default");
        file.delete();
        Files.writeString(file.toPath(), content, UTF_8);
    }

    private static void renameJs(final File file) {
        file.renameTo(new File(file.getParentFile(), file.getName().replace(".js", ".ts")));
    }

}
````

</details>

- [x] Rename files JS to TS
- [x] Regenerate index.ts files + fix exports
- [x] Add type definitions to index.ts files
- [x] ~~Fix type errors~~ (Reverted as requested in https://github.com/faker-js/faker/pull/238#issuecomment-1017137290)
- [x] Avoid type errors
- [x] Verify the data structure hasn't changed at all

I recommend reviewing the PR commit by commit instead of all at once.

**Tests that nothing has been changed**

1) Checkout the main branch and run:

````ts
import * as index from "@faker-js/faker/src/locales";
console.log(JSON.stringify(index));
````

via `pnpm esno index.ts > output_old.json`

2) Then checkout this PR and it again

via `pnpm esno index.ts > output_new.json`

3) Sort both jsons by keys.
4) Compare output_old.json with output_new.json
5) Verify that output_new does not remove or rename any values and only adds new ones

**Accidental fixes**

- https://github.com/faker-js/faker/blob/66aa374fc3ca958070819d7eb6b119accc01cd3d/src/locales/hr/index.ts#L9
- nb_NO: I renamed the femine and masculine first name files to match the other languages, but I also had to change the name template because it uses these files.
